### PR TITLE
Add success and message fields to IO services

### PR DESCRIPTION
**Pull request by steviedale**
This pull request addresses issue ros-industrial#362, adding success and message fields to the Motoman IO services

### CODE REVIEW

**Comment by steviedale** [comment]
@gavanderhoorn I created this branch off of the `steviedale:update/separate_io_from_joint_streamer` branch (from this [PR](https://github.com/ros-industrial/motoman/pull/365)). I figured this would avoid merge conflicts, but if this is against best practice I can rebase off of the `ros-industrial:kinetic-devel` branch instead.

**Comment by gavanderhoorn** [comment]
You'll probably want to `rebase` this after we merge #365.

**Comment by gavanderhoorn** [comment]
Could you clarify why you added c80c351? `std::to_string(..)` looks like it makes sense where you used it.

**Comment by marip8** [comment]
A couple general comments:

- I don't find it useful to print error messages returned by services to the console. If you call a service from the command line this message will be printed automatically when the service completes, and if a caller of the service wants to log the error then they can do so in the calling code. To me, this additional logging just adds more code and clutters the terminal output unnecessarily. I would recommend removing the ROS logging from the service callbacks
- I prefer to make make service callback functions always return true and return the logical success in the service response instead. When a service returns false, it's ambiguous whether the service was unreachable (which also causes `ros::service::call(...)` to return false) or if there was a logical failure.

**Comment by gavanderhoorn** [comment]
>  * I don't find it useful to print error messages returned by services to the console. If you call a service from the command line this message will be printed automatically when the service completes, and if a caller of the service wants to log the error then they can do so in the calling code. To me, this additional logging just adds more code and clutters the terminal output unnecessarily. I would recommend removing the ROS logging from the service callbacks

~Are you thinking of ROS services?~

~Afaict, @steviedale kept the logging which was already there in the callbacks of the Simple Message service.~

~It seems like it would be informative to keep those, as it allows to quickly see whether at least the Simple Message part of the communication succeeded or not.~

I misread the diff.

I'm not against printing the message to the console, as it's a nice way first step in debugging or diagnosing a "why does it not work" situation. It's for humans only though, so @marip8 is right in that callers could do this themselves.

> * I prefer to make make service callback functions always return true and return the logical success in the service response instead. When a service returns false, it's ambiguous whether the service was unreachable (which also causes `ros::service::call(...)` to return false) or if there was a logical failure.

Assuming you are referring to ROS services here: there are very specific constraints on how services should behave: [roscpp/Overview/Services: Callback Signature](http://wiki.ros.org/roscpp/Overview/Services#Callback_Signature).

`false` should be returned when the server failed to process the service request.

So we should probably be returning `true` here always, as long as were able to forward the request to the MotoROS side. Then communicate success of the read/write operation using the `result` and `message` fields.

That was the rationale for adding the `result` and `message` fields (also to `std_srvs/Trigger`), as the `bool` return from the callback should not be used for this.

**Comment by steviedale** [comment]
> Could you clarify why you added [c80c351](https://github.com/ros-industrial/motoman/commit/c80c351138beaff915f7b1b555d899aafcae99eb)? `std::to_string(..)` looks like it makes sense where you used it.

I changed this because the kinetic CI was failing (the melodic passes though). It looks like maybe the version of C++ being used for kinetic does not have std::to_string. I believe this function was added in C++11. I'm not sure what version the kinetic CI is set up for or if it can be easily changed, but I figured an easy solution would be to use a more backwards compatible method for int to string conversion.

**Comment by steviedale** [comment]
Apologies for the force commit and any confusion that might bring. I messed up a rebase and needed to fix it. The changes are the same as before with the addition of always returning `true` for io service calls.

**Comment by gavanderhoorn** [comment]
@steviedale: with #365 merged I believe this PR contains some duplicated commits?

**Comment by steviedale** [comment]
@gavanderhoorn I just fixed the commit history. Are there any changes that still need to be addressed for this pull request?

File: motoman_driver/src/io_relay.cpp [hunk 1]
```diff
@@ -113,12 +114,20 @@ bool MotomanIORelay::readSingleIoCB(
 
   if (!result)
   {
-    ROS_ERROR_NAMED("io.read", "Reading IO element %d failed", req.address);
-    return false;
+    std::stringstream message;
+    message << "Reading IO element " << req.address << " failed";
```
Comment (marip8) [inline motoman_driver/src/io_relay.cpp]: Please update the messages here with a single quote or parentheses around the streamed variables like `req.address`. In the case that a user specifies a blank address, this error message would be somewhat confusing and unhelpful: `Reading IO element  failed` vs `Reading IO element () failed`

Comment (gavanderhoorn) [reply motoman_driver/src/io_relay.cpp]: Isn't `address` an integer?

Comment (marip8) [reply motoman_driver/src/io_relay.cpp]:
```suggestion
    message << "Reading IO element (" << req.address << ") failed";
```

Comment (gavanderhoorn) [reply motoman_driver/src/io_relay.cpp]: @marip8: if `address` is an integer, can it ever be empty?

Comment (marip8) [reply motoman_driver/src/io_relay.cpp]: Good point; I would think it probably couldn't be empty (opposed to strings that are blank by default). This change is probably superfluous then

Comment (gavanderhoorn) [reply motoman_driver/src/io_relay.cpp]: I fully agree with you in case this was a `string` btw. Nothing more perplexing than seeing weird error messages.

**Comment by gavanderhoorn** [comment]
@steviedale: I've added some commits. Some cosmetic, but I also wanted to avoid having to pass `char*` to `ROS_ERROR_NAMED(..)` (as we do actually have a `std::string`), and we don't need to add a "success message" to the service response. `success` is already `true`, so it's clear things worked.

Perhaps the last change also connects to what @marip8 wrote in his earlier comment.

The `message` is there to help the caller (ie: human in this case still) determine what happened. But in case of success, we already know, so no need for the additional message.

**Comment by gavanderhoorn** [comment]
@steviedale: could you let me know whether you are OK with my additions?

I've tested them with our YRC1000, and on invalid writes I now get:

```
$ rosservice call /write_single_io "{ address: 10010, value: 2 }"
success: False
message: "Writing to IO element 10010 failed"
```

instead of `service returned an error:` (and then nothing, due to use returning `false`).

**Comment by steviedale** [comment]
@gavanderhoorn I just looked over your changes, looks good to me.

**Review by gavanderhoorn — APPROVED** [review]
Thanks @steviedale for the contribution :+1:

**Pull request merged**